### DISCUSSION TITLE
Issue with plotsurvfitDS after using ds.Surv and ds.survfit

### DISCUSSION THREAD

**Discussion by naudinlo** [neelsoumya/dsSurvivalClient#12] [Q&A] [answered]
**Describe the bug**
Hello, when using `dsSurvivalClient::ds.Surv` and `dsSurvivalClient::ds.survfit` I have an issue. I cannot plot the output with `plotsurvfitDS` as it is an object and not a vector, making the `ds.plotsurvfit.R` code failing at line 119.

**To Reproduce**
```
ds.Surv(time="D_patients$dob_day", time2="D_patients$dod_day", "D_patients$expire_flag", type="counting", objectname='surv_object')
ds.survfit(formula='surv_object~1', objectname='survfit_object')
ds.plotsurvfit(formula='surv_object')
```

Generates the error:

`[1] "Command 'plotsurvfitDS(formula = \"surv_object\", NULL)' failed on 'abc': Error while evaluating 'dsSurvival::plotsurvfitDS(formula = \"surv_object\", NULL)' -> Error : $ operator is invalid for atomic vectors\n"`

**Comment by neelsoumya** [comment]
can you please give some more details like what VM are you using, what data are you using and what versions of Opal and R you have? Also can you reproduce this error using synthetic data? Then I can try to reproduce this error message.

**Comment by naudinlo** [comment]
Thanks for your answer. 

On the client side, we use the rocker/rstudio:4.1.2 docker image which has Ubuntu 20.04.5 LTS and R version 4.1.2.
For opal, we use the obiba/opal:4.5 image where we installed dsSurvival version 2.1.0 (we also tried to use the demo server at https://opal-demo.obiba.org/ with the same result).
We tried to use the data that can be found in the SURVIVAL project on https://opal-demo.obiba.org/.
```
builder <- DSI::newDSLoginBuilder(.silent = TRUE)
builder$append(
 server = "chu",
 url = "https://opal-demo.obiba.org/",
 user = "administrator", password = "password",
 driver = "OpalDriver",
 options='list(ssl_verifyhost=0, ssl_verifypeer=0)',
 profile="survival"
)
logindata <- builder$build()
connections <- DSI::datashield.login(logins = logindata)
datashield.assign.table(connections, symbol = "D", table = "SURVIVAL.EXPAND_WITH_MISSING1")
ds.asNumeric(x.name = "D$cens", newobj = "EVENT", datasources = connections)
ds.asNumeric(x.name = "D$starttime", newobj = "STARTTIME", datasources = connections)
ds.asNumeric(x.name = "D$endtime", newobj = "ENDTIME", datasources = connections)
ds.Surv(time="STARTTIME", time2="ENDTIME", event="EVENT", type="counting", objectname='surv_object')
ds.survfit(formula='surv_object~1', objectname='survfit_object')
ds.plotsurvfit(formula = 'survfit_object')
```
The last line fails with the following error:
```
Error in xy.coords(x, y, xlabel, ylabel, log) :
 'x' is a list, but does not have components 'x' and ‘y’
```

 
 Thanks,

**Comment by neelsoumya** [comment]
Thanks @naudinlo 

Could you please try this using just R and R Studio and just client-side package installations?
This uses the demo server at

https://opal-sandbox.mrc-epid.cam.ac.uk

The following steps work for me and I can generate the survival curves.

1. Could you please follow this simple script:

https://github.com/neelsoumya/dsSurvival/blob/main/vignettes/simple_script.R

This works with just client side installations i.e. R and R Studio

2. I installed all the required packages using this script:

https://raw.githubusercontent.com/neelsoumya/dsSurvival/main/installer_R_Rstudio.R

3. My output from sessionInfo() shows the following:

R version 4.2.1 (2022-06-23)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Monterey 12.4

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib

locale:
[1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] dsBaseClient_6.1.1     DSOpal_1.4.0           opalr_3.2.0            httr_1.4.3            
 [5] dsSurvivalClient_2.1.0 DSI_1.5.0              R6_2.5.1               progress_1.2.2        
 [9] ggplot2_3.3.6          metafor_3.4-0          metadat_1.2-0          Matrix_1.4-1          
[13] survival_3.4-0         tinytex_0.41           rmarkdown_2.15         knitr_1.39

**Comment by neelsoumya** [comment]
@naudinlo @tombisho 

I also tried your code and ran it on my laptop (Mac OS X) and R and R Studio (no docker image). This works fine. Here is the code below (I modified the code slightly by loading the libraries):

```r


###################
# load libraries
###################
library(survival)
library(metafor)
library(dsSurvivalClient)
require('DSI')
require('DSOpal')
require('dsBaseClient')


builder <- DSI::newDSLoginBuilder(.silent = TRUE)
builder$append(
  server = "chu",
  url = "https://opal-demo.obiba.org/",
  user = "administrator", password = "password",
  driver = "OpalDriver",
  options='list(ssl_verifyhost=0, ssl_verifypeer=0)',
  profile="survival"
)
logindata <- builder$build()
connections <- DSI::datashield.login(logins = logindata)
datashield.assign.table(connections, symbol = "D", table = "SURVIVAL.EXPAND_WITH_MISSING1")
ds.asNumeric(x.name = "D$cens", newobj = "EVENT", datasources = connections)
ds.asNumeric(x.name = "D$starttime", newobj = "STARTTIME", datasources = connections)
ds.asNumeric(x.name = "D$endtime", newobj = "ENDTIME", datasources = connections)
ds.Surv(time="STARTTIME", time2="ENDTIME", event="EVENT", type="counting", objectname='surv_object')
ds.survfit(formula='surv_object~1', objectname='survfit_object')
ds.plotsurvfit(formula = 'survfit_object')


sessionInfo()

```

> sessionInfo()
R version 4.2.1 (2022-06-23)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Monterey 12.4

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib

locale:
[1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] dsBaseClient_6.1.1     DSOpal_1.4.0           opalr_3.2.0            httr_1.4.3            
 [5] dsSurvivalClient_2.1.0 DSI_1.5.0              R6_2.5.1               progress_1.2.2        
 [9] metafor_3.4-0          metadat_1.2-0          Matrix_1.4-1           survival_3.4-0

**Comment by neelsoumya** [comment, marked as answer]
@naudinlo @tombisho  I tried all this code on my laptop on R and R Studio. I did not use the docker image. Can you please try using this setup and let me know if it works for you? 

i.e. do the following

 1. Install R and R Studio on your laptop
 
 2. Install all the required packages using these commands in R or R Studio:
 
```r

install.packages('devtools')
library(devtools)
devtools::install_github('neelsoumya/dsSurvivalClient')
devtools::install_github('datashield/dsBaseClient@6.1.1')
install.packages('metafor')
install.packages('DSOpal')
install.packages('DSI')
install.packages('opalr')
install.packages('meta')
install.packages('fANCOVA')
install.packages('survival')

```

 3. Run the following script


```r



###################
# load libraries
###################
library(survival)
library(metafor)
library(dsSurvivalClient)
require('DSI')
require('DSOpal')
require('dsBaseClient')

builder <- DSI::newDSLoginBuilder(.silent = TRUE)
builder$append(
  server = "chu",
  url = "https://opal-demo.obiba.org/",
  user = "administrator", password = "password",
  driver = "OpalDriver",
  options='list(ssl_verifyhost=0, ssl_verifypeer=0)',
  profile="survival"
)

logindata <- builder$build()
connections <- DSI::datashield.login(logins = logindata)

datashield.assign.table(connections, symbol = "D", table = "SURVIVAL.EXPAND_WITH_MISSING1")

ds.asNumeric(x.name = "D$cens", newobj = "EVENT", datasources = connections)
ds.asNumeric(x.name = "D$starttime", newobj = "STARTTIME", datasources = connections)
ds.asNumeric(x.name = "D$endtime", newobj = "ENDTIME", datasources = connections)
ds.Surv(time="STARTTIME", time2="ENDTIME", event="EVENT", type="counting", objectname='surv_object')
ds.survfit(formula='surv_object~1', objectname='survfit_object')
ds.plotsurvfit(formula = 'survfit_object')


sessionInfo()


```

**Comment by neelsoumya** [comment]
@naudinlo did this code work for you?

**Reply by neelsoumya** [reply]
@naudinlo did this help fix your problem?
